lib/dialect/src/ConstraintsAnalysis.cpp

```diff
@@ -0,0 +1,1229 @@
+	}
+
+	// Method for either inserting a value or joining the existing one 
+	bool ConstraintsLattice::insertOrJoin(const std::pair<mlir::Value,bool>& key, const ConstraintsLattice::IntegerRange& other){
+		auto found=this->ranges.find(key);
+		if(found!=this->ranges.end()){
+			auto copy=found->second;
+			found->second= 
+				ConstraintsLattice::joinRange(found->second, other);
+			return copy!=found->second;
+		}
+		else{
+			this->ranges.insert(std::make_pair(key, other));
+			return true;
+		}
+	}
+
+	// Method for either inserting a value or meeting the existing one 
+	bool ConstraintsLattice::insertOrMeet(const std::pair<mlir::Value,bool>& key, const ConstraintsLattice::IntegerRange& other){
+		auto found=this->ranges.find(key);
+		if(found!=this->ranges.end()){
+			auto copy=found->second;
+			found->second= 
+				ConstraintsLattice::meetRange(found->second, other);
+			return copy!=found->second;
+		}
+		else{
+			this->ranges.insert(std::make_pair(key, other));
+			return true;
+		}
+	}
+
+	// Method for either inserting a value or updating the existing one 
+	bool ConstraintsLattice::insertOrUpdate(const std::pair<mlir::Value,bool>& key, const ConstraintsLattice::IntegerRange& other){
+		auto found=this->ranges.find(key);
+		if(found!=this->ranges.end()){
+			auto copy=found->second;
+			found->second=other;
+			return copy!=found->second;
+		}
+		else{
+			this->ranges.insert(
+				std::make_pair(key, other));
+			return true;
+		}
+	}
+
+	// Method for either inserting a value or joining the existing one (overload)
+	bool ConstraintsLattice::insertOrUpdate(const std::pair<mlir::Value,bool>& keyToInsert, const std::optional<Integer>& min, const std::optional<Integer>& max){
+		auto found=this->ranges.find(keyToInsert);
+		if(found!=this->ranges.end()){
+			auto copy=found->second;
+			found->second=ConstraintsLattice::updateRange(found->second, min, max);
+			return copy!=found->second;
+		}
+		else{
+			this->ranges.insert(
+				std::make_pair(keyToInsert,
+					createRange(
+						min.value_or(MIN),
+						max.value_or(MAX))));
+			return true;
+		}
+	}
+
+	// Method for either inserting a value in the to_check structure
+	void ConstraintsLattice::insertInTocheck(const mlir::Value& value){
+		this->to_check.insert(value);
+	}
+
+	// This function is actually quite simpler since we always know 
+	// the lhs is not a constant and the rhs is a constant
+	// Result is useful to understand which check do to the operation
+	bool ConstraintsLattice::updateRelationalRange(mlir::Operation* statement, bool result, const mlir::Value& to_update, const mlir::Operation* other, const bool branch_to_update){
+
+		auto other_casted=mlir::dyn_cast<mlir::rlc::Constant>(other);
+
+		auto rhs_const=other_casted->getAttr("value").dyn_cast<IntegerAttr>().getInt();
+
+		std::optional<Integer> min;
+		std::optional<Integer> max;
+
+		llvm::TypeSwitch<mlir::Operation*>(statement)
+		.Case<mlir::rlc::GreaterOp>([&](mlir::rlc::GreaterOp op){
```

Comment:
constraint lattice cannot know the exhaustive list of operations to handle here, because they may change.